File: cpp_mandelbrot/mandelbrot_benchmark.cpp
#include <iostream>
#include <vector>
#include <complex>
#include <chrono>
#include <thread>
#include <fstream>
#include <iomanip>
#include <mutex>
#include <string>
#include <algorithm>
#include <functional>
#include <cmath>

// Define a structure for benchmark results
struct BenchmarkResult
{
    std::string region_name;
    double single_thread_time;
    double best_time;
    int best_threads;
    double speedup;
};

// Define a struct for region parameters
struct Region
{
    double y_min, y_max, x_min, x_max;
    std::string name;
    int extra_iterations; // Additional iterations for complex regions
};

// Define calculation methods
enum class CalcMethod
{
    STANDARD,     // Standard z = z^2 + c
    BURNING_SHIP, // Burning ship fractal
    MULTIBROT     // Higher power Multibrot set
};

// Function to calculate the Mandelbrot set for a single point with standard formula
int mandelbrot_point(std::complex<double> c, int max_iterations)
{
    std::complex<double> z(0, 0);
    int iterations = 0;

    while (std::abs(z) <= 2.0 && iterations < max_iterations)
    {
        z = z * z + c;
        iterations++;
    }

    // Smooth coloring algorithm (more CPU intensive)
    if (iterations < max_iterations)
    {
        // Add smoothing for more precise color transitions
        double log_zn = log(std::abs(z));
        double nu = log(log_zn / log(2.0)) / log(2.0);
        iterations = iterations + 1 - nu;
    }

    return iterations;
}

// Function to calculate the Burning Ship fractal (more compute-intensive variant)
int burning_ship_point(std::complex<double> c, int max_iterations)
{
    double real = 0;
    double imag = 0;
    int iterations = 0;

    while (real * real + imag * imag <= 4.0 && iterations < max_iterations)
    {
        double next_real = real * real - imag * imag + c.real();
        double next_imag = 2.0 * std::abs(real * imag) + c.imag();
        real = next_real;
        imag = next_imag;
        iterations++;
    }

    // Smooth coloring
    if (iterations < max_iterations)
    {
        double zn = sqrt(real * real + imag * imag);
        double nu = log(log(zn)) / log(2.0);
        iterations = iterations + 1 - nu;
    }

    return iterations;
}

// Function to calculate the Multibrot set (higher power z^n + c, more CPU intensive)
int multibrot_point(std::complex<double> c, int max_iterations, int power = 4)
{
    std::complex<double> z(0, 0);
    int iterations = 0;

    while (std::abs(z) <= 2.0 && iterations < max_iterations)
    {
        // Compute z^power (more expensive than z^2)
        std::complex<double> z_powered = z;
        for (int i = 1; i < power; i++)
        {
            z_powered *= z;
        }
        z = z_powered + c;
        iterations++;
    }

    // Smooth coloring
    if (iterations < max_iterations)
    {
        double log_zn = log(std::abs(z));
        double nu = log(log_zn / log(2.0)) / log(2.0);
        iterations = iterations + 1 - nu;
    }

    return iterations;
}

// Function to calculate a point using the selected method
int calculate_point(std::complex<double> c, int max_iterations, CalcMethod method, int power = 4)
{
    switch (method)
    {
    case CalcMethod::BURNING_SHIP:
        return burning_ship_point(c, max_iterations);
    case CalcMethod::MULTIBROT:
        return multibrot_point(c, max_iterations, power);
    case CalcMethod::STANDARD:
    default:
        return mandelbrot_point(c, max_iterations);
    }
}

// Function to calculate the Mandelbrot set for a range of points (single-threaded)
std::vector<std::vector<int>> mandelbrot(int height, int width, int max_iterations,
                                         double y_min, double y_max, double x_min, double x_max,
                                         CalcMethod method = CalcMethod::STANDARD)
{
    std::vector<std::vector<int>> result(height, std::vector<int>(width));

    for (int y = 0; y < height; y++)
    {
        for (int x = 0; x < width; x++)
        {
            double real = x_min + (x_max - x_min) * x / width;
            double imag = y_min + (y_max - y_min) * y / height;
            std::complex<double> c(real, imag);
            result[y][x] = calculate_point(c, max_iterations, method);
        }
    }

    return result;
}

// Function to calculate a chunk of the Mandelbrot set (for parallel processing)
void mandelbrot_chunk(int y_start, int y_end, int width, int max_iterations,
                      double y_min, double y_max, double x_min, double x_max,
                      std::vector<std::vector<int>> &result,
                      CalcMethod method = CalcMethod::STANDARD)
{
    for (int y = y_start; y < y_end; y++)
    {
        for (int x = 0; x < width; x++)
        {
            double real = x_min + (x_max - x_min) * x / width;
            double imag = y_min + (y_max - y_min) * y / (result.size());
            std::complex<double> c(real, imag);
            result[y][x] = calculate_point(c, max_iterations, method);
        }
    }
}

// Function to calculate the Mandelbrot set using multiple threads
std::vector<std::vector<int>> mandelbrot_parallel(int height, int width, int max_iterations,
                                                  double y_min, double y_max, double x_min, double x_max,
                                                  int num_threads,
                                                  CalcMethod method = CalcMethod::STANDARD)
{
    std::vector<std::vector<int>> result(height, std::vector<int>(width));
    std::vector<std::thread> threads;

    // Calculate chunk size for each thread
    int chunk_size = height / num_threads;

    // Start threads
    for (int i = 0; i < num_threads; i++)
    {
        int y_start = i * chunk_size;
        int y_end = (i == num_threads - 1) ? height : (i + 1) * chunk_size;

        threads.push_back(std::thread(mandelbrot_chunk, y_start, y_end, width, max_iterations,
                                      y_min, y_max, x_min, x_max, std::ref(result), method));
    }

    // Join threads
    for (auto &t : threads)
    {
        t.join();
    }

    return result;
}

// Function to save the Mandelbrot set as a PPM image file
void save_image(const std::vector<std::vector<int>> &data, int max_iterations,
                const std::string &filename)
{
    int height = data.size();
    int width = data[0].size();

    std::ofstream file(filename, std::ios::binary);
    file << "P6\n"
         << width << " " << height << "\n255\n";

    for (int y = 0; y < height; y++)
    {
        for (int x = 0; x < width; x++)
        {
            double value = data[y][x];

            // More complex color calculation (more CPU intensive)
            unsigned char r, g, b;
            if (value >= max_iterations)
            {
                r = g = b = 0; // Black for points in the set
            }
            else
            {
                // Enhanced color gradient for points outside the set
                double normalized = value / max_iterations;

                // More complex HSV to RGB conversion (more computationally intensive)
                double h = 360.0 * normalized;
                double s = 1.0;
                double v = normalized < 0.5 ? normalized * 2 : 1.0;

                // HSV to RGB conversion
                double c = v * s;
                double x = c * (1 - std::abs(std::fmod(h / 60.0, 2) - 1));
                double m = v - c;

                double r1, g1, b1;
                if (h < 60)
                {
                    r1 = c;
                    g1 = x;
                    b1 = 0;
                }
                else if (h < 120)
                {
                    r1 = x;
                    g1 = c;
                    b1 = 0;
                }
                else if (h < 180)
                {
                    r1 = 0;
                    g1 = c;
                    b1 = x;
                }
                else if (h < 240)
                {
                    r1 = 0;
                    g1 = x;
                    b1 = c;
                }
                else if (h < 300)
                {
                    r1 = x;
                    g1 = 0;
                    b1 = c;
                }
                else
                {
                    r1 = c;
                    g1 = 0;
                    b1 = x;
                }

                r = static_cast<unsigned char>((r1 + m) * 255);
                g = static_cast<unsigned char>((g1 + m) * 255);
                b = static_cast<unsigned char>((b1 + m) * 255);
            }

            file.write(reinterpret_cast<char *>(&r), 1);
            file.write(reinterpret_cast<char *>(&g), 1);
            file.write(reinterpret_cast<char *>(&b), 1);
        }
    }

    file.close();
}

// String representation of calculation method
std::string method_to_string(CalcMethod method)
{
    switch (method)
    {
    case CalcMethod::BURNING_SHIP:
        return "Burning Ship";
    case CalcMethod::MULTIBROT:
        return "Multibrot";
    case CalcMethod::STANDARD:
        return "Standard";
    default:
        return "Unknown";
    }
}

// Function to run benchmarks with different numbers of threads
BenchmarkResult benchmark(int height, int width, int max_iterations,
                          double y_min, double y_max, double x_min, double x_max,
                          const std::string &region_name,
                          CalcMethod method = CalcMethod::STANDARD)
{
    std::cout << "\nBenchmarking region: " << region_name << " (" << method_to_string(method) << ")" << std::endl;

    // Run single-threaded version
    std::cout << "Running single-threaded benchmark..." << std::endl;
    auto start = std::chrono::high_resolution_clock::now();
    mandelbrot(height, width, max_iterations, y_min, y_max, x_min, x_max, method);
    auto end = std::chrono::high_resolution_clock::now();
    double single_thread_time = std::chrono::duration<double>(end - start).count();
    std::cout << "Single-threaded time: " << single_thread_time << " seconds" << std::endl;

    // Run multithreaded versions with different thread counts
    std::vector<int> thread_counts = {2, 4, 8, std::max(10, static_cast<int>(std::thread::hardware_concurrency()))};
    std::vector<double> times;

    for (int num_threads : thread_counts)
    {
        std::cout << "Running with " << num_threads << " threads..." << std::endl;
        start = std::chrono::high_resolution_clock::now();
        mandelbrot_parallel(height, width, max_iterations, y_min, y_max, x_min, x_max, num_threads, method);
        end = std::chrono::high_resolution_clock::now();
        double thread_time = std::chrono::duration<double>(end - start).count();
        times.push_back(thread_time);
        std::cout << num_threads << " threads: " << thread_time << " seconds, speedup: "
                  << single_thread_time / thread_time << "x" << std::endl;
    }

    // Find the best performance
    auto min_it = std::min_element(times.begin(), times.end());
    double best_time = *min_it;
    int best_threads = thread_counts[std::distance(times.begin(), min_it)];
    double speedup = single_thread_time / best_time;

    return {region_name + " (" + method_to_string(method) + ")", single_thread_time, best_time, best_threads, speedup};
}

int main()
{
    // Parameters - increased resolution and iterations for an even more challenging benchmark
    int height = 4000;         // 2x the original resolution
    int width = 5000;          // 2x the original resolution
    int max_iterations = 2000; // 4x the original iteration count

    // Define interesting regions to explore with extra iterations for complex areas
    std::vector<Region> regions = {
        // Full view
        {-1.25, 1.25, -2.0, 0.5, "full", 0},
        // Deep zoom to a complex area (Seahorse Valley)
        {-0.745, -0.735, 0.21, 0.22, "seahorse", 500},
        // Feigenbaum point - very computationally intensive
        {-1.401155, -1.401150, 0.0000, 0.0000045, "feigenbaum", 1000},
        // Mini Mandelbrot deep zoom - extremely computationally intensive
        {-1.77777515, -1.77777505, 0.0117, 0.0118, "mini", 2000},
        // Spiral structure with very fine detail
        {-0.747203, -0.747196, -0.1151, -0.1150, "spiral", 1500}};

    // Define calculation methods to benchmark (increasing complexity)
    std::vector<CalcMethod> methods = {
        CalcMethod::STANDARD,     // Original Mandelbrot
        CalcMethod::BURNING_SHIP, // Burning Ship (more CPU intensive)
        CalcMethod::MULTIBROT     // Multibrot with higher powers (most CPU intensive)
    };

    std::vector<BenchmarkResult> all_results;

    for (const auto &method : methods)
    {
        for (const auto &region : regions)
        {
            // Adjust max iterations based on region complexity
            int iterations = max_iterations + region.extra_iterations;

            // Run benchmark (skip some combinations to save time)
            if (method == CalcMethod::MULTIBROT &&
                (region.name == "feigenbaum" || region.name == "mini"))
            {
                std::cout << "Skipping extremely intensive " << method_to_string(method)
                          << " calculation for " << region.name << std::endl;
                continue;
            }

            BenchmarkResult result = benchmark(height, width, iterations,
                                               region.y_min, region.y_max, region.x_min, region.x_max,
                                               region.name, method);
            all_results.push_back(result);

            // Generate the final image using all available cores
            if (method == CalcMethod::STANDARD)
            { // Only save images for standard method
                int cpu_count = std::thread::hardware_concurrency();
                std::cout << "\nGenerating final image using " << cpu_count << " threads..." << std::endl;
                auto start = std::chrono::high_resolution_clock::now();
                auto image = mandelbrot_parallel(height, width, iterations,
                                                 region.y_min, region.y_max, region.x_min, region.x_max,
                                                 cpu_count, method);
                auto end = std::chrono::high_resolution_clock::now();
                std::cout << "Image generation time: "
                          << std::chrono::duration<double>(end - start).count() << " seconds" << std::endl;

                // Save the image
                std::string filename = "mandelbrot_" + region.name + ".ppm";
                save_image(image, iterations, filename);
                std::cout << "Image saved as " << filename << std::endl;
            }
        }
    }

    // Print summary
    std::cout << "\n"
              << std::string(70, '=') << std::endl;
    std::cout << "BENCHMARK SUMMARY" << std::endl;
    std::cout << std::string(70, '=') << std::endl;
    std::cout << std::left << std::setw(25) << "Region"
              << std::setw(12) << "Single (s)"
              << std::setw(12) << "Best (s)"
              << std::setw(10) << "Threads"
              << std::setw(10) << "Speedup" << std::endl;
    std::cout << std::string(70, '-') << std::endl;

    double total_speedup = 0.0;
    for (const auto &result : all_results)
    {
        std::cout << std::left << std::setw(25) << result.region_name
                  << std::fixed << std::setprecision(3)
                  << std::setw(12) << result.single_thread_time
                  << std::setw(12) << result.best_time
                  << std::setw(10) << result.best_threads
                  << std::setw(10) << std::setprecision(2) << result.speedup << "x" << std::endl;
        total_speedup += result.speedup;
    }

    double avg_speedup = total_speedup / all_results.size();
    std::cout << std::string(70, '-') << std::endl;
    std::cout << "Average speedup across all regions: " << std::fixed << std::setprecision(2)
              << avg_speedup << "x" << std::endl;
    std::cout << "CPU count: " << std::thread::hardware_concurrency() << std::endl;
    std::cout << "Total computation: " << height << "x" << width << " pixels × "
              << "variable max iterations × " << regions.size() << " regions × "
              << methods.size() << " methods" << std::endl;
    std::cout << std::string(70, '=') << std::endl;

    return 0;
}

File: mandelbrot_benchmark.py
#!/usr/bin/env python3
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import LinearSegmentedColormap
import time
import concurrent.futures
import multiprocessing

def mandelbrot(h, w, max_iters, y_min, y_max, x_min, x_max):
    """
    Calculate the Mandelbrot set for the given region and resolution.
    
    Args:
        h, w: Height and width of the resulting image
        max_iters: Maximum number of iterations for each point
        y_min, y_max, x_min, x_max: Boundaries of the region to calculate
    
    Returns:
        2D numpy array of iteration counts
    """
    y, x = np.ogrid[y_min:y_max:h*1j, x_min:x_max:w*1j]
    c = x + y*1j
    z = c
    divtime = max_iters + np.zeros(z.shape, dtype=int)
    
    for i in range(max_iters):
        z = z**2 + c
        diverge = z*np.conj(z) > 2**2
        div_now = diverge & (divtime == max_iters)
        divtime[div_now] = i
        z[diverge] = 2
        
    return divtime

def mandelbrot_chunk(args):
    """Process a horizontal chunk of the Mandelbrot image"""
    y_start, y_end, h_chunk, w, max_iters, y_min, y_max, x_min, x_max = args
    
    # Calculate the y range for this chunk
    y_chunk_min = y_min + (y_max - y_min) * y_start / h_chunk
    y_chunk_max = y_min + (y_max - y_min) * y_end / h_chunk
    
    return mandelbrot(y_end - y_start, w, max_iters, y_chunk_min, y_chunk_max, x_min, x_max)

def mandelbrot_parallel(h, w, max_iters, y_min, y_max, x_min, x_max, num_threads=None):
    """Calculate the Mandelbrot set using multiple threads"""
    if num_threads is None:
        num_threads = multiprocessing.cpu_count()
    
    # Split the image into horizontal chunks
    chunk_size = h // num_threads
    chunks = []
    
    for i in range(num_threads):
        y_start = i * chunk_size
        y_end = (i + 1) * chunk_size if i < num_threads - 1 else h
        chunks.append((y_start, y_end, h, w, max_iters, y_min, y_max, x_min, x_max))
    
    # Process chunks in parallel
    with concurrent.futures.ThreadPoolExecutor(max_workers=num_threads) as executor:
        results = list(executor.map(mandelbrot_chunk, chunks))
    
    # Combine the results
    combined = np.vstack(results)
    return combined

def benchmark(h, w, max_iters, y_min, y_max, x_min, x_max):
    """Run benchmarks with different numbers of threads"""
    
    # First run a single-threaded version
    print("Running single-threaded benchmark...")
    start = time.time()
    mandelbrot(h, w, max_iters, y_min, y_max, x_min, x_max)
    single_thread_time = time.time() - start
    print(f"Single-threaded time: {single_thread_time:.3f} seconds")
    
    thread_counts = [2, 4, 8, max(10, multiprocessing.cpu_count())]
    times = []
    
    for num_threads in thread_counts:
        print(f"Running with {num_threads} threads...")
        start = time.time()
        mandelbrot_parallel(h, w, max_iters, y_min, y_max, x_min, x_max, num_threads)
        thread_time = time.time() - start
        times.append(thread_time)
        print(f"{num_threads} threads: {thread_time:.3f} seconds, speedup: {single_thread_time/thread_time:.2f}x")
    
    return single_thread_time, thread_counts, times

def visualize_mandelbrot(image, h, w, y_min, y_max, x_min, x_max, filename=None):
    """Create a visualization of the Mandelbrot set"""
    # Create a custom colormap - black for points in the set, colored gradient for outside
    colors = [(0, 0, 0), (0, 0, 0.5), (0, 0.5, 1), (1, 1, 1), (1, 0.5, 0), (0.5, 0, 0), (0, 0, 0)]
    cmap = LinearSegmentedColormap.from_list('mandelbrot', colors, N=256)
    
    plt.figure(figsize=(12, 10))
    plt.imshow(image, cmap=cmap, origin='lower', extent=(x_min, x_max, y_min, y_max))
    plt.title(f'Mandelbrot Set ({h}x{w}, region: [{x_min}, {x_max}] x [{y_min}, {y_max}])')
    plt.colorbar(label='Iterations')
    
    if filename:
        plt.savefig(filename, dpi=150)
    else:
        plt.show()

def main():
    # Parameters - increased resolution and iterations for a more demanding benchmark
    h, w = 4000, 5000  # Image height and width (doubled from original)
    max_iters = 500     # Maximum iterations (5x more than original)
    
    # Interesting regions to explore - added more challenging regions
    regions = [
        # Full view
        {'y_min': -1.25, 'y_max': 1.25, 'x_min': -2.0, 'x_max': 0.5, 'name': 'full'},
        # Zoomed in area 
        {'y_min': -0.15, 'y_max': 0.15, 'x_min': -0.8, 'x_max': -0.5, 'name': 'zoom1'},
        # Deep zoom to a complex area (Seahorse Valley) - computationally intensive
        {'y_min': -0.745, 'y_max': -0.735, 'x_min': 0.21, 'x_max': 0.22, 'name': 'seahorse'},
        # Spiral structure - complex area requiring many iterations
        {'y_min': -0.7472, 'y_max': -0.7468, 'x_min': -0.1152, 'x_max': -0.1148, 'name': 'spiral'}
    ]
    
    all_results = []
    
    for region in regions:
        print(f"\nBenchmarking region: {region['name']}")
        y_min, y_max = region['y_min'], region['y_max']
        x_min, x_max = region['x_min'], region['x_max']
        
        # Run benchmark
        single_time, thread_counts, times = benchmark(h, w, max_iters, y_min, y_max, x_min, x_max)
        
        # Generate the final image using all available cores
        cpu_count = multiprocessing.cpu_count()
        print(f"\nGenerating final image using {cpu_count} threads...")
        start = time.time()
        image = mandelbrot_parallel(h, w, max_iters, y_min, y_max, x_min, x_max, cpu_count)
        print(f"Image generation time: {time.time() - start:.3f} seconds")
        
        # Save the image
        filename = f"mandelbrot_{region['name']}.png"
        visualize_mandelbrot(image, h, w, y_min, y_max, x_min, x_max, filename)
        print(f"Image saved as {filename}")
        
        # Plot benchmark results
        plt.figure(figsize=(10, 6))
        plt.plot([1] + thread_counts, [single_time] + times, 'o-')
        plt.axhline(y=single_time, color='r', linestyle='--', alpha=0.3)
        plt.title(f'Mandelbrot Benchmark - Region: {region["name"]}')
        plt.xlabel('Number of threads')
        plt.ylabel('Time (seconds)')
        plt.grid(True, alpha=0.3)
        for i, (threads, time_val) in enumerate(zip([1] + thread_counts, [single_time] + times)):
            plt.annotate(f'{time_val:.2f}s', (threads, time_val), 
                         textcoords="offset points", xytext=(0,10), ha='center')
        plt.savefig(f"benchmark_{region['name']}.png")
        
        # Store results for summary
        best_time = min(times)
        best_threads = thread_counts[times.index(best_time)]
        speedup = single_time / best_time
        all_results.append({
            'region': region['name'],
            'single_thread': single_time,
            'best_time': best_time,
            'best_threads': best_threads,
            'speedup': speedup
        })
    
    # Print summary
    print("\n" + "="*60)
    print("BENCHMARK SUMMARY")
    print("="*60)
    print(f"{'Region':<10} {'Single (s)':<12} {'Best (s)':<12} {'Threads':<10} {'Speedup':<10}")
    print("-"*60)
    
    total_speedup = 0
    for result in all_results:
        print(f"{result['region']:<10} {result['single_thread']:<12.3f} {result['best_time']:<12.3f} {result['best_threads']:<10} {result['speedup']:<10.2f}x")
        total_speedup += result['speedup']
    
    avg_speedup = total_speedup / len(all_results)
    print("-"*60)
    print(f"Average speedup across all regions: {avg_speedup:.2f}x")
    print(f"CPU count: {multiprocessing.cpu_count()}")
    print(f"Total computation: {h}x{w} pixels × {max_iters} max iterations × {len(regions)} regions")
    print("="*60)

if __name__ == "__main__":
    main() 